memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+  | AAELF64_ | IHI 0056               | ELF for the Arm 64-bit Architecture |
+  +----------+------------------------+-------------------------------------+
+
+Terms and Abbreviations
+-----------------------
+
+The MemtagABI extension for the Arm 64-bit Architecture uses the following terms
+and abbreviations.
+
+AArch64
+   The 64-bit general-purpose register width state of the Armv8 architecture.
+
+ABI
+   Application Binary Interface:
+
+   1. The specifications to which an executable must conform in order to
+      execute in a specific execution environment. For example, the
+      *Linux ABI for the Arm Architecture*.
+
+   2. A particular aspect of the specifications to which independently produced
+      relocatable files must conform in order to be statically linkable and
+      executable.  For example, the CPPABI64_, AAELF64_, ...
+
+.. raw:: pdf
+
+   PageBreak
+
+Scope
+=====
+
+This document is a set of extensions to ELF for the Arm 64-bit architecture
+(AAELF64_) describing how the MemtagABI information is encoded in the ELF file.
+As an alpha document all details in this document are subject to change.
+
+.. raw:: pdf
+
+   PageBreak
+
+Platform Standards
+==================
+
+As is the case with the AAELF64_, we expect that each operating system that
+adopts components of this ABI specification will specify additional requirements
+and constraints that must be met by application code in binary form and the
+code-generation tools that generate such code. This document will present
+recommendations for a SysVr4 like operating system such as Linux.
+
+.. raw:: pdf
+
+   PageBreak
+
+Introduction
+============
+
+The Armv8.5-A architecture introduced the Memory Tagging Extensions (MTE)
+feature, which allows each 16-byte region (a "tag granule") of taggable memory
+to have a 4-bit piece of metadata (a "memory tag") associated with it.
+Additionally, the 2nd top nibble of a pointer may contain a similar piece of
+metadata (a "pointer tag" or "allocation tag"). During load and store
+instructions, the pointer tag and memory tag are compared, and upon mismatch,
+the processor aborts with a Synchronous or Asynchronous MTE exception.
+
+Design Goals
+------------
+
+The goals of the MemtagABIELF64 document are to:
+
+- Provide primitives that can be used to support protection of the heap, stack,
+  and global variables with memory tags.
+
+- Provide rationale for design choices.
+
+- Describe a scheme for tagging of the heap and global variables that is
+  backwards-compatible. In particular, a binary built with MTE heap or globals
+  protection will function correctly when linked with a non-MemtagABI aware
+  linker, loaded with a non-MemtagABI aware dynamic loader, or run on non-MTE
+  capable hardware. Note, it is expected that stack instrumentation, because MTE
+  instructions are not part of the no-op space, may crash on non-MTE hardware or
+  non-MemtagABI linkers/loaders.
```

Comment:
> For the non-MemTagABI loaders, is it that the loader wouldn't setup the stack in tagged memory and then when the store tag instructions run it will crash?

Almost. The stack won't be tagged, and so the inline `stg`-family instructions generated by the compiler instrumentation will be no-ops. Basically, on MTE HW with a non-MemtagABI-aware loader, stack instrumentation doesn't crash but doesn't do anything interesting.

Reply:
Ok right, I just made the assumption it would fault but the pseudo code backs you up. Makes sense now.